site/topics/searching/searching.rst

```diff
@@ -54,32 +54,77 @@ Searching (again)
-   
-    If you can, code up the solution. 
-   
+* This is an example of a *binary search*
+* Like linear search, binary search is used to find an element within some collection
+* However, if I was doing a linear search for a number between :math:`0 -- 1023`, I could only guarantee that I would find the element in :math:`1024` guesses
+* But with binary search, I was able to do it in :math:`10` or less guesses
+* Though, this required the higher/lower information --- the data was sorted
 
-* Now we need to ask the same questions as before:
-    * On average, how many iterations through your loop does your function make?
-    * How about in the worst case?
-    * Is your solution the best possible?
-    * Might there exist some super clever algorithm that is somehow better (faster) than yours?	
-			
-* This is a *very* common pattern in developing algorithms:
-    * The more *general* your problem is, the slower the solution is.
-    * The more you know about the *structure* of your problem (e.g., "the list is always sorted"), the more opportunities you have to use that knowledge to make the solution faster.
+    * A linear search has no such requirement
+
+
+Complexity Analysis
+-------------------
+
+* The magic with binary search is that, with every guess I made, I was able to eliminate half of the remaining numbers
+
+    * My first guess was :math:`512` --- if you said *lower* I know the number is between :math:`0 -- 511`, if you said *higher* I know it's between :math:`513 -- 1023`
+
+* To generalize the idea, if I had :math:`n` numbers, and I guess the number :math:`\frac{n}{2}`
+
+    * If you say lower, then the number must be between :math:`0 -- \frac{n}{2} - 1`
+    * If you said higher, then the number must be between :math:`\frac{n}{2} + 1 -- (n - 1)`
+
+* With linear search, every guess only eliminated one number
+
+* Once again, let's consider the worse case scenario --- the ``needle`` is not within the ``haystack`` of size :math:`1024`
+
+    * One guess gets me to :math:`512` numbers
+    * Two guesses gets me to :math:`256` numbers
+    * Three guesses get me to :math:`128`
+    * Four gets me to :math:`64`
+    * Five gets me to :math:`32`
+    * Six gets me :math:`16`
+    * Seven :math:`8`
+    * Eight :math:`4`
+    * Nine :math:`2`
+    * Ten :math:`1`
+
+
+* Originally, with linear search, the relationship between the input :math:`n` and the amount of work is :math:`n`
+
+    * If there are :math:`n` things in the ``haystack``, I have to look at all :math:`n`
+    * If we doubled the size to :math:`2n`, the amount of things needed to be looked at also doubles to :math:`2n`
+
+* With binary search however, the relationship between the size of the input :math:`n` and the amount of work is :math:`log_{2}(n)`
+
+    * Doubling the size to :math:`2n` only adds one more guess
+
+        * :math:`log_{2}(2n)`
+        * :math:`log_{2}(n) + log_{2}(2)`
+        * :math:`log_{2}(n) + 1`
```

Comment:
This is probably unnecessary. Intuitively, doubling the size adds only one step to the progression above that starts at 1024.

Reply:
Agreed. Additionally, if anything, including it is worse:
- They see more logs and freak our
- They think the logs are important
- They don't see the forest through the trees